strawberry/schema/execute.py

```diff
@@ -63,17 +80,27 @@ async def execute(
                 extensions=await extensions_runner.get_extensions_results(),
             )
 
-        if validate_queries:
-            async with extensions_runner.validation():
-                validation_errors = validate(schema, document, rules=validation_rules)
-
-            if validation_errors:
-                execution_context.errors = validation_errors
-                return ExecutionResult(data=None, errors=validation_errors)
+        async with extensions_runner.validation():
+            # Check if there are any validation rules or if validation has
+            # already been run by an extension
+            if (
+                len(execution_context.validation_rules) > 0
+                and execution_context.errors is None
+            ):
+                assert execution_context.graphql_schema
+                assert execution_context.graphql_document
+                execution_context.errors = validate_document(
+                    execution_context.graphql_schema,
+                    execution_context.graphql_document,
+                    execution_context.validation_rules,
+                )
+
+            if execution_context.errors:
+                return ExecutionResult(data=None, errors=execution_context.errors)
```

Comment:
there's a bit of code duplication now, we should find a way to deduplicate it, can't think of any solution at the moment though

Reply:
Yeah I was thinking of just wrapping this up into a function but I was struggling with a good name for it...